packages/core/src/validate.js

```diff
@@ -264,17 +265,25 @@ export default function validateFormData(
 }
 
 /**
- * Get a similar schema where ref's are prefixed with "__rjsf_rootSchema"
+ * Get a similar schema where ref's are prefixed with `rootSchemaId`
  * This is used in isValid to make references to the rootSchema
  */
-function withIdRefPrefix(schema) {
-  const obj = { ...schema };
-  for (let key of Object.keys(obj)) {
-    const value = obj[key];
-    if (key === "$ref") {
-      obj[key] = "__rjsf_rootSchema" + value;
-    } else if (value.constructor === Object) {
-      obj[key] = withIdRefPrefix(value);
+function withIdRefPrefix(schemaNode) {
+  let obj = schemaNode;
+  if (schemaNode.constructor === Object) {
+    obj = { ...schemaNode };
+    for (let key of Object.keys(obj)) {
+      const value = obj[key];
+      if (key === "$ref") {
+        obj[key] = rootSchemaId + value;
+      } else {
+        obj[key] = withIdRefPrefix(value);
+      }
+    }
+  } else if (Array.isArray(schemaNode)) {
+    obj = [...schemaNode];
+    for (var i = 0; i < obj.length; i++) {
+      obj[i] = withIdRefPrefix(obj[i]);
```

Comment:
What about a schema like this:

```json
{
  "title": "A registration form",
  "description": "A simple form example.",
  "type": "object",
  "properties": {
    "$ref": {
      "type": "string",
      "title": "First name",
      "default": "Chuck"
    }
  }
}
```

It _is_ a valid schema (see [playground example](https://rjsf-team.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6eyIkcmVmIjoiQ2h1Y2siLCJmaXJzdE5hbWUiOiJDaHVjayIsImxhc3ROYW1lIjoiTm9ycmlzIiwiYWdlIjo3NSwiYmlvIjoiUm91bmRob3VzZSBraWNraW5nIGFzc2VzIHNpbmNlIDE5NDAiLCJwYXNzd29yZCI6Im5vbmVlZCJ9LCJzY2hlbWEiOnsidGl0bGUiOiJBIHJlZ2lzdHJhdGlvbiBmb3JtIiwiZGVzY3JpcHRpb24iOiJBIHNpbXBsZSBmb3JtIGV4YW1wbGUuIiwidHlwZSI6Im9iamVjdCIsInByb3BlcnRpZXMiOnsiJHJlZiI6eyJ0eXBlIjoic3RyaW5nIiwidGl0bGUiOiJGaXJzdCBuYW1lIiwiZGVmYXVsdCI6IkNodWNrIn19fSwidWlTY2hlbWEiOnsiZmlyc3ROYW1lIjp7InVpOmF1dG9mb2N1cyI6dHJ1ZSwidWk6ZW1wdHlWYWx1ZSI6IiIsInVpOmF1dG9jb21wbGV0ZSI6ImZhbWlseS1uYW1lIn0sImxhc3ROYW1lIjp7InVpOmVtcHR5VmFsdWUiOiIiLCJ1aTphdXRvY29tcGxldGUiOiJnaXZlbi1uYW1lIn0sImFnZSI6eyJ1aTp3aWRnZXQiOiJ1cGRvd24iLCJ1aTp0aXRsZSI6IkFnZSBvZiBwZXJzb24iLCJ1aTpkZXNjcmlwdGlvbiI6IihlYXJ0aGlhbiB5ZWFyKSJ9LCJiaW8iOnsidWk6d2lkZ2V0IjoidGV4dGFyZWEifSwicGFzc3dvcmQiOnsidWk6d2lkZ2V0IjoicGFzc3dvcmQiLCJ1aTpoZWxwIjoiSGludDogTWFrZSBpdCBzdHJvbmchIn0sImRhdGUiOnsidWk6d2lkZ2V0IjoiYWx0LWRhdGV0aW1lIn0sInRlbGVwaG9uZSI6eyJ1aTpvcHRpb25zIjp7ImlucHV0VHlwZSI6InRlbCJ9fX0sInRoZW1lIjoiZGVmYXVsdCIsImxpdmVTZXR0aW5ncyI6eyJ2YWxpZGF0ZSI6ZmFsc2UsImRpc2FibGUiOmZhbHNlLCJvbWl0RXh0cmFEYXRhIjpmYWxzZSwibGl2ZU9taXQiOmZhbHNlfX0=)).

I'm not sure if it would be easy to fix this use case, though. Let me know what you think.

Reply:
Maybe we could just check to ensure that the ref's value is a string, and we rename the ref only in that case?